### PR TITLE
Implement NSIndexSet enumeration and *PassingTest apis

### LINKED ISSUE #2076: Implement NSIndexSet index* apis
indexPassingTest:
indexWithOptions:passingTest:
indexInRange:options:passingTest:

indexesPassingTest:
indexesWithOptions:passingTest:
indexesInRange:options:passingTest:

getIndexes:maxCount:inIndexRange:

### LINKED ISSUE #2077: Implement NSIndexSet enumerate* apis
enumerateRangesInRange:options:usingBlock:
enumerateRangesUsingBlock:
enumerateRangesWithOptions:usingBlock:
enumerateIndexesInRange:options:usingBlock:

### PR DESCRIPTION
**Pull request by aballway**
Fixes #2076 
Fixes #2077

### CODE REVIEW

File: Frameworks/Foundation/NSIndexSet.mm [hunk 1]
```diff
@@ -350,72 +314,109 @@ - (BOOL)isEqualToIndexSet:(NSIndexSet*)otherSet {
...
 
 /**
- @Status Stub
+ @Status Interoperable
 */
-- (NSUInteger)indexInRange:(NSRange)range options:(NSEnumerationOptions)opts passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (NSUInteger)indexInRange:(NSRange)range options:(NSEnumerationOptions)options passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate {
+    NSIndexSet* set = [self indexesInRange:range options:options passingTest:predicate];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSIndexSet.mm]: `*stop` optimization?

File: Frameworks/Foundation/NSIndexSet.mm [hunk 2]
```diff
@@ -350,72 +314,109 @@ - (BOOL)isEqualToIndexSet:(NSIndexSet*)otherSet {
...
 }
 
 /**
- @Status Stub
+ @Status Interoperable
 */
-- (void)enumerateRangesInRange:(NSRange)range options:(NSEnumerationOptions)opts usingBlock:(void (^)(NSRange, BOOL*))block {
-    UNIMPLEMENTED();
+- (void)enumerateRangesUsingBlock:(void (^)(NSRange, BOOL*))block {
+    [self enumerateRangesWithOptions:0 usingBlock:block];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSIndexSet.mm]: please always call the most-specific version in these shim implementations; you'll avoid an entire level of dynamic dispatch and a stack frame.

File: Frameworks/Foundation/NSIndexSet.mm [hunk 3]
```diff
@@ -483,18 +484,78 @@ - (NSUInteger)indexGreaterThanOrEqualToIndex:(NSUInteger)index {
...
+    [self enumerateRangesInRange:enumerationRange
+                         options:0
+                      usingBlock:^(NSRange range, BOOL* stop) {
+                          NSUInteger j = range.location;
+                          for (; j < range.location + range.length && index < bufferSize; ++j, ++index) {
+                              indexBuffer[index] = j;
+                          }
+
+                          if (indexRange) {
+                              *indexRange = { j, range.location + range.length - j };
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSIndexSet.mm]: huh; indexRange is in/out?

File: include/Foundation/NSIndexSet.h [hunk 4]
```diff
@@ -35,26 +35,24 @@ FOUNDATION_EXPORT_CLASS
...
-- (void)enumerateRangesInRange:(NSRange)range options:(NSEnumerationOptions)opts usingBlock:(void (^)(NSRange, BOOL*))block STUB_METHOD;
-- (void)enumerateRangesUsingBlock:(void (^)(NSRange, BOOL*))block STUB_METHOD;
-- (void)enumerateRangesWithOptions:(NSEnumerationOptions)opts usingBlock:(void (^)(NSRange, BOOL*))block STUB_METHOD;
+- (NSUInteger)indexPassingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (NSIndexSet*)indexesPassingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (NSUInteger)indexWithOptions:(NSEnumerationOptions)opts passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (NSIndexSet*)indexesWithOptions:(NSEnumerationOptions)opts passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (NSUInteger)indexInRange:(NSRange)range options:(NSEnumerationOptions)opts passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (NSIndexSet*)indexesInRange:(NSRange)range options:(NSEnumerationOptions)opts passingTest:(BOOL (^)(NSUInteger, BOOL*))predicate;
+- (void)enumerateRangesInRange:(NSRange)range options:(NSEnumerationOptions)opts usingBlock:(void (^)(NSRange, BOOL*))block;
```
Comment (memontic-ms) [inline include/Foundation/NSIndexSet.h]: >opts [](start = 76, length = 4)

nit: Update var names to reflect changes in NSIndexSet.mm

File: Frameworks/Foundation/NSIndexSet.mm [hunk 5]
```diff
@@ -483,18 +484,78 @@ - (NSUInteger)indexGreaterThanOrEqualToIndex:(NSUInteger)index {
...
+    NSUInteger count = [static_cast<NSMutableIndexSet*>(self) _count];
+    NSRangePointer ranges = [static_cast<NSMutableIndexSet*>(self) _allRanges];
+    NSUInteger start = (reverse) ? count : 1;
+    NSUInteger end = (reverse) ? 0 : count + 1;
+    NSInteger step = (reverse) ? -1 : 1;
+    __block BOOL stop = NO;
+    for (NSUInteger i = start; i != end && stop == NO; i += step) {
+        NSRange intersection = NSIntersectionRange(range, ranges[i - 1]);
+        if (intersection.length != 0) {
+            NSUInteger rangeStart = reverse ? intersection.location + intersection.length : intersection.location + 1;
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSIndexSet.mm]: you're right, this DOES seem like it should be able to be lifted out/commonized.

Comment (aballway) [inline Frameworks/Foundation/NSIndexSet.mm]: IKR. Unfortunately, in the concurrent case we'd be doing two levels of dispatch_group_async which seems sillier than very similar but not identical code

File: Frameworks/Foundation/NSIndexSet.mm (hunk 3 above)
Comment (aballway) [inline Frameworks/Foundation/NSIndexSet.mm]: yup :falkland_islands:

File: Frameworks/Foundation/NSIndexSet.mm [hunk 6]
```diff
@@ -345,7 +345,16 @@ - (NSIndexSet*)indexesWithOptions:(NSEnumerationOptions)options passingTest:(BOO
...
+                                   options:options
+                               passingTest:^BOOL(NSUInteger index, BOOL* stop) {
+                                   BOOL ret = predicate(index, stop);
+                                   if (ret == YES) {
+                                       *stop = YES;
+                                   }
+
+                                   return ret;
+                               }];
     return (options & NSEnumerationReverse) ? [set lastIndex] : [set firstIndex];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSIndexSet.mm]: since set will contain only one index, this is no longer necessary

**Comment by DHowett-MSFT** [comment]
nit: because this is a set, the set will set set

**Pull request merged**